pkg/services/oauthtoken/oauth_token.go

```diff
@@ -126,76 +135,98 @@ func (o *Service) HasOAuthEntry(ctx context.Context, usr identity.Requester) (*l
 }
 
 // TryTokenRefresh returns an error in case the OAuth token refresh was unsuccessful
-// It uses a singleflight.Group to prevent getting the Refresh Token multiple times for a given User
-func (o *Service) TryTokenRefresh(ctx context.Context, usr identity.Requester) error {
+// It uses a server lock to prevent getting the Refresh Token multiple times for a given User
+func (o *Service) TryTokenRefresh(ctx context.Context, usr identity.Requester) (*oauth2.Token, error) {
 	ctxLogger := logger.FromContext(ctx)
 
 	if usr == nil || usr.IsNil() {
 		ctxLogger.Warn("Can only refresh OAuth tokens for existing users", "user", "nil")
 		// Not user, no token.
-		return nil
+		return nil, nil
 	}
 
 	if !usr.IsIdentityType(claims.TypeUser) {
 		ctxLogger.Warn("Can only refresh OAuth tokens for users", "id", usr.GetID())
-		return nil
+		return nil, nil
 	}
 
 	userID, err := usr.GetInternalID()
 	if err != nil {
 		ctxLogger.Warn("Failed to convert user id to int", "id", usr.GetID(), "error", err)
-		return nil
+		return nil, nil
 	}
 
 	ctxLogger = ctxLogger.New("userID", userID)
 
+	// get the token's auth provider (f.e. azuread)
+	currAuthenticator := usr.GetAuthenticatedBy()
+	if !strings.HasPrefix(currAuthenticator, "oauth") {
+		ctxLogger.Warn("The specified user's auth provider is not OAuth", "authmodule", currAuthenticator)
```

Comment:
This should be a debug log. As far as I can tell till will be called for everything that has been authenticated with a session

Reply:
Either that or this check should be moved to the hook

Reply:
>  As far as I can tell till will be called for everything that has been authenticated with a session
`SyncAuthTokenHook` has an extra check for OAuth before it calls the `TryTokenRefresh` function + `GetCurrentOAuthToken` (which is another call site of `TryTokenRefresh`) has also have the extra check. I'm not against changing this to Debug, but at least with the Warn it will be  pretty obvious if the function is called from somewhere (in the future) which doesn't have the check.

Reply:
Yes you are correct. we already check if they are authenticate with oauth there. Leaving it as a warning log is fine then :)